cmds/cmd.c

```diff
@@ -28,61 +29,53 @@
 
 
 /* Linker symbol points to the beginning of .data section */
-extern char script[];
+extern const char script[];
 
 struct {
 	const cmd_t *cmds[SIZE_CMDS];
-	char args[MAX_CMD_ARGS_NB][SIZE_CMD_ARG_LINE];
-
 	int ll;
 	int cl;
 	char lines[SIZE_HIST][SIZE_CMD_ARG_LINE];
 } cmd_common;
 
 
-static void cmd_skipblanks(const char *line, unsigned int *pos, const char *blanks)
+static int cmd_parseArgLine(const char **lines, char *buf, size_t bufsz, char *argv[], size_t argvsz)
 {
-	const char *p;
+	char *const end = buf + bufsz;
+	int argc = 0;
 
-	while (line[*pos]) {
-		for (p = blanks; *p && line[*pos] != *p; p++);
-		if (!*p)
-			break;
-		(*pos)++;
-	}
-}
+	if (lines == NULL || *lines == NULL || **lines == '\0')
+		return -1; /* EOF */
 
+	while (**lines != '\0') {
+		if (isspace(**lines)) {
+			if (isblank(*(*lines)++))
+				continue;
+			else
+				break;
+		}
 
-/* TODO: old code needs to be cleaned up */
-static char *cmd_getnext(const char *line, unsigned int *pos, const char *blanks, char *word, unsigned int len)
-{
-	char c;
-	unsigned int i, wp = 0;
+		if (argc + 1 >= argvsz) {
```

Comment:
This condition allows to have one argument less than the maximum size is defined. For example, `map` should consists of 5 arguments: `map itcm 0x0 0x40000 rwx`. When I have set `MAX_CMD_ARGS_NB = 5`, the parsing has stopped on `map` command.

Reply:
Yes, `MAX_CMD_ARGS_NB` shall be  increased by `+1`, this is due to the argument lis termination with `NULL`, this is from standard C, where You could count arguments till `NULL`. So, I'll increase by  +1 `MAX_CMD_ARGS_NB` marking that last argument `+1` is `NULL`, as we talked.
```
#define MAX_CMD_ARGS_NB (10 + 1)
```

Reply:
Ok. Can you write a comment in the code why `+1` is added to avoid doubts in the future?

Reply:
Isn't it simpler to do it this way:

cmd.h
```
#define MAX_CMD_ARGS_NB 10
```
cmd.c
```
int cmd_parse(const char *script)
{
	char *argv[MAX_CMD_ARGS_NB + 1]; // and even +2 (cmd nane + NULL)

		argc = cmd_parseArgLine(&script, argline, SIZE_CMD_ARG_LINE, argv, sizeof(argv) / sizeof(argv[0]);

```

Reply:
@HBuczynski , changed name of MAX_CMD_ARGS_NB to SIZE_CMD_ARGV to denote array size, also added comments where apropriate.
@kemonats +1, +2 .. is not the case. We should be sure of the length of the array everywhere, so `cmd_parseArgLine()` checks `(argc + 1) > SIZE_CMD_ARGV` , thus `argv` should be at least one for terminating `NULL` pointer, then for this edge case there will be no space for arguments only terminating `NULL` pointer will be there - thats valid case. And. If `SIZE_CMD_ARGV` is zero then `cmd_parseArgLine()` should return negative `EINVAL`.

Reply:
In cmd_parseArgLine() checks (argc + 1) > argvsz
```
static int cmd_parseArgLine(const char **lines, char *buf, size_t bufsz, char *argv[], size_t argvsz)

argc = cmd_parseArgLine(&script, argline, SIZE_CMD_ARG_LINE, argv, sizeof(argv) / sizeof(argv[0]);
```
and so argvsz equals the number of elements of the argv array.

Reply:
Exactly. `SIZE_CMD_ARGV` tells how many pointers are in argv[], but the last pointer is reserved and always points to  `NULL` , thus +1 to indicate that fact, see comments there. It `(argc + 1) > argvsz` is not actual, see merged version:
https://github.com/phoenix-rtos/plo/blob/d9fefe2330d9aa771825db235a1826565534b864/cmds/cmd.c#L58-L60